composer/callbacks/memory_snapshot.py

```diff
@@ -0,0 +1,161 @@
+
+            Whenever a trace file is saved, it is also uploaded as a file according to this format string.
+            The same format variables as for ``filename`` are available.
+
+            .. seealso:: :doc:`Uploading Files</trainer/file_uploading>` for notes for file uploading.
+
+            Leading slashes (``'/'``) will be stripped.
+
+            To disable uploading trace files, set this parameter to ``None``.
+        overwrite (bool, optional): Whether to override existing memory snapshots. Defaults to False.
+
+            If False, then the trace folder as determined by ``folder`` must be empty.
+    """
+
+    def __init__(
+        self,
+        skip_batches: int = 1,
+        interval: Union[int, str, Time] = '3ba',
+        max_entries: int = 100000,
+        folder: str = '{run_name}/torch_traces',
+        filename: str = 'rank{rank}.{batch}.pickle',
+        remote_file_name: Optional[str] = '{run_name}/torch_traces/rank{rank}.{batch}.pickle',
+        overwrite: bool = False,
+    ) -> None:
+        self.batches_left_to_skip = skip_batches
+        # Check that the interval timestring is parsable and convert into time object
+        self.interval = Time.from_input(interval, TimeUnit.BATCH)
+        self.max_entries = max_entries
+        self.folder = folder
+        self.folder_name = None
+        self.filename = filename
+        self.remote_file_name = remote_file_name
+        self.overwrite = overwrite
+        self._enabled = True
+        self._start_time = None
+
+    def init(self, state: State, logger: Logger) -> None:
+        # Not relying on `torch.cuda.is_available()` since the model could be on CPU.
+        model_device = next(state.model.parameters()).device
+
+        if model_device.type not in ('cuda', 'meta'):
+            warnings.warn(f'The memory snapshot only works on CUDA devices, but the model is on {model_device.type}.')
+
+        self.folder_name = format_name_with_dist(self.folder, state.run_name)
+        os.makedirs(self.folder_name, exist_ok=True)
+        if not self.overwrite:
+            ensure_folder_is_empty(self.folder_name)
+
+    def batch_start(self, state: State, logger: Logger) -> None:
+        if self._enabled and self._start_time is None and self.batches_left_to_skip == 0:
+            self.start_record_memory_history()
+            self._start_time = state.timestamp.get(self.interval.unit).value
+
+    def batch_end(self, state: State, logger: Logger) -> None:
+        if not self._enabled:
+            return
+
+        if self.batches_left_to_skip > 0:
+            self.batches_left_to_skip -= 1
+            return
+        assert self._start_time is not None
+
+        if state.timestamp.get(self.interval.unit).value == (self._start_time + self.interval.value):
+            self.export_memory_snapshot(state, logger)
+            self.stop_record_memory_history()
+            self._start_time = None
+            self._enabled = False
```

Comment:
Does this mean we will only record the cuda memory profile once? If so, can we record this in the docstring. 

A design decision to be made is should we incorporate this logic as a callback or into the existing torch profiler?

I'm okay with having it as a callback because it takes less memory/time but it is a bit confusing to the user that some memory traces are stored as profile flags and others are stored as callbacks.